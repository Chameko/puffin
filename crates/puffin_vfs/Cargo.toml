[package]
name = "puffin_vfs"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
crossbeam-channel = "0.5.8"
fxhash = "0.2"
notify = "5.1.0"
relative-path = { version = "1.8", default-features = false }
thiserror = "1.0"
