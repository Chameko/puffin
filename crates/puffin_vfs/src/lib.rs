/// Contains the path interner
mod path_interner;
use path_interner::PathInterner;

/// Contains the moniter
mod monitor;
pub use monitor::Monitor;
pub use monitor::ForegroundMessage;
pub use monitor::MonitorMessage;

/// Contains the path types used in Puffin
pub mod path;
pub use path::AbsPath;
pub use path::AbsPathBuf;

use relative_path::RelativePath;
use thiserror::Error;

/// Identifier for a file in the virtual file system
#[derive(Debug, Clone, Copy, PartialEq, PartialOrd, Eq, Ord)]
#[repr(transparent)]
pub struct FileID(pub u32);

/// Error type for the virtual file system
#[derive(Error, Debug)]
pub enum VFSError {
    #[error("expected an absolute path")]
    NotAbsolutePath,
}

/// The virtual file system
#[derive(Debug)]
pub struct VFS {
    /// The internal path interner
    interner: PathInterner,
}

impl VFS {
    /// Either adds a path and returns the [`FileID`] or returns the [`FileID`] if it already exists
    pub fn intern(&mut self, path: &AbsPath) -> FileID {
        self.interner.intern(path)
    }

    /// Get an absolute path from a file ID
    pub fn get_path(&self, id: FileID) -> Option<&AbsPath> {
        self.interner.get_path(id)
    }

    /// Intern a [RelativePath] that is relative to a supplied [AbsPath]
    pub fn intern_relative_path(&mut self, path: &RelativePath, relative_to: &AbsPath) -> FileID {
        self.interner.intern(&AbsPathBuf::try_from(path.to_logical_path(relative_to)).unwrap())
    }
}
