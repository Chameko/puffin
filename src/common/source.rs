use ropey::{Rope, RopeSlice};
use std::ops::RangeBounds;
use std::{io::BufReader, path::PathBuf};

use crate::diagnostic::PuffinError;

/// A file in the source
#[derive(Debug)]
pub struct File {
    pub text: Rope,
    path: PathBuf,
}

impl File {
    /// Create a new file from a path
    pub fn new(path: &str) -> Result<Self, PuffinError> {
        let file = std::fs::File::open(path)?;
        let rope = Rope::from_reader(BufReader::new(file)).expect("Unable to create rope");
        Ok(Self {
            text: rope,
            path: path.into(),
        })
    }

    /// Get a slice of the text
    pub fn get_slice<R: RangeBounds<usize> + Clone>(&self, range: &R) -> RopeSlice {
        self.text
            .get_slice(range.clone())
            .expect("Expected in bounds range")
    }

    /// Get a slice of the text as a string
    pub fn get_string<R: RangeBounds<usize> + Clone>(&self, range: &R) -> String {
        self.get_slice(range).to_string()
    }

    /// Get the path as a string
    pub fn display_path(&self) -> String {
        format!("{}", self.path.display())
    }

    /// Get the text as a vec of chars
    pub fn get_chars(&self) -> Vec<char> {
        self.text.chars().collect()
    }
}

/// The whole source as a list of files
pub struct Source {
    /// Files in the source
    pub files: Vec<File>,
}

impl Source {
    /// Create a new source
    pub fn new(path: &str) -> Result<Self, PuffinError> {
        let file = File::new(path)?;
        Ok(Self { files: vec![file] })
    }
}
