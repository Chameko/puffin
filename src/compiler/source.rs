use crate::diagnostic::PetrelError;
use std::fs::File;
use std::io::Read;

/// The petrel source
#[derive(Debug)]
pub struct Source {
    pub src: String,
}

#[derive(Clone, Debug, PartialEq, Eq)]
/// A convenient way of getting a slice from the source
pub struct Span {
    pub start: usize,
    pub end: usize,
}

impl Source {
    pub fn from_file(path: &str) -> Result<Self, PetrelError> {
        // Read file
        let mut file = File::open(path)?;
        let mut input: String = "".into();
        file.read_to_string(&mut input)?;

        // Create an iterator of said characters.
        // This is done as chars() references input, which is a local variable
        Ok(Self { src: input })
    }

    pub fn slice(&self, span: &Span) -> Option<&str> {
        self.src.get(span.start..span.end)
    }
}

impl Span {
    pub fn new(start: usize, length: usize) -> Self {
        Self {
            start,
            end: start + length,
        }
    }
}
