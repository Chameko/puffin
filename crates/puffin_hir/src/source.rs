use puffin_vfs::{AbsPathBuf, FileID};
use relative_path::RelativePathBuf;
use std::sync::Arc;

/// Type use for marking slices of text in the source files
pub type TextSlice = std::ops::RangeInclusive<usize>;

/// The query group for interacting with the source files
#[salsa::query_group(SourceStorage)]
pub trait SourceDatabase {
    /// Set the input file
    #[salsa::input]
    fn input_src(&self) -> Arc<Source>;

    /// Get the source tree
    #[salsa::input]
    fn source_tree(&self) -> Arc<SourceTree>;
}

/// The source tree represents a tree of sources in a Puffin project. Each file is relative to the source tree
#[derive(Debug, Clone)]
pub struct SourceTree {
    /// The base directory of the source tree. Everything else is relative to this
    dir: FileID,
    /// The sources in the source tree
    sources: Vec<Source>
}

impl SourceTree {
    pub fn find_source(&self, file_id: FileID) -> Option<&Source> {
        for source in &self.sources {
            if source.file == file_id {
                return Some(&source)
            }
        }
        None
    }
}

/// A source in the source tree
#[derive(Debug, Clone, PartialEq)]
pub struct Source {
    // The file that the source belongs to
    pub file: FileID,
    // The text contained within the file
    pub text: String,
}

impl Source {
    /// Get the text of a source file
    pub fn get_text(&self) -> &str {
        self.text.as_str()
    }
}
