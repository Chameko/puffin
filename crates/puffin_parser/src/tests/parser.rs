use puffin_hir::source::Source;
use puffin_vfs::FileID;
use crate::{TokenStream, lexer::Lexer, parser::Parser};
use super::output_cst;

/// Parses the inputed string and creates a snapshot
fn standard_test(src: &str) -> String {
    let lexer = Lexer::new(src);
    let source = Source{ file: FileID(0), text: src.to_string() };
    let parser = Parser::new(TokenStream::new(lexer.start_scan()), source);
    let parse = parser.parse();
    if parse.errors.len() > 0 {
        for error in parse.errors {
            println!("{}", error.debug_display("test.pf", src.split_inclusive('\n').enumerate().map(|mut v| {
                v.0 += 1;
                v
            }).collect()));
            panic!("Errors detected when parsing");
        }
    }
    output_cst(&parse.green_node)
}

#[test]
fn just_number() {
    insta::assert_snapshot!(standard_test("1"));
}

#[test]
fn simple_expr() {
    insta::assert_snapshot!(standard_test("1 + 2"));
}

#[test]
fn multiple_simple() {
    insta::assert_snapshot!(standard_test("1 + 2 + 3 + 4"));
}

#[test]
fn prefix_operation() {
    insta::assert_snapshot!(standard_test("-1 + 2"));
}

#[test]
fn multiple_prefix_operation() {
    insta::assert_snapshot!(standard_test("--1 + -2"));
}

#[test]
fn order_of_operations() {
    insta::assert_snapshot!(standard_test("1 + 2 * 5 - 3"));
}

#[test]
fn paren() {
    insta::assert_snapshot!(standard_test("1 + 2 * (3 - 2) + (1 * 2)"));
}

#[test]
fn paren_2() {
    insta::assert_snapshot!(standard_test("(1 + 2 + 3)"));
}

#[test]
fn paren_3() {
    insta::assert_snapshot!(standard_test("(1 + (1)) - (2 + 1 + 2)"));
}

#[test]
fn print_stmt() {
    insta::assert_snapshot!(standard_test("print 2 * (4 - 2)"));
}

#[test]
fn multiple_stmt() {
    insta::assert_snapshot!(standard_test("1 + 2\nprint 3\n2 * 4"));
}

#[test]
fn let_stmt() {
    insta::assert_snapshot!(standard_test("let banana = 1"));
}

#[test]
fn block_stmt() {
    insta::assert_snapshot!(standard_test("{ print 1 + 1 }"));
}

#[test]
fn block_stmt_2() {
    insta::assert_snapshot!(standard_test("{\nprint 1 + 1\n}"));
}

#[test]
fn block_stmt_3() {
    insta::assert_snapshot!(standard_test("{print 1 + 1\n}"));
}

#[test]
fn block_stmt_4() {
    insta::assert_snapshot!(standard_test("{\nprint 1 + 1}"));
}

#[test]
fn assign_stmt() {
    insta::assert_snapshot!(standard_test("a = 4 + 3"));
}
