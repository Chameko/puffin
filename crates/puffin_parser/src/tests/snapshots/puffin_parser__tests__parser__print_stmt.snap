---
source: crates/puffin_parser/src/tests/parser.rs
expression: "standard_test(\"print 2 * (4 - 2)\")"
---
SOURCE_FILE@0..=16
    PRINT_STMT@0..=16
        print |KW_PRINT@0..4
          |WHITESPACE@5..5
        BIN_EXPR@6..=16
            PAT_EXPR@6..=6
                LIT_PAT@6..=6
                    2 |INT@6..6
              |WHITESPACE@7..7
            * |STAR@8..8
              |WHITESPACE@9..9
            PAREN_EXPR@10..=16
                ( |L_PAREN@10..10
                BIN_EXPR@11..=15
                    PAT_EXPR@11..=11
                        LIT_PAT@11..=11
                            4 |INT@11..11
                      |WHITESPACE@12..12
                    - |MINUS@13..13
                      |WHITESPACE@14..14
                    PAT_EXPR@15..=15
                        LIT_PAT@15..=15
                            2 |INT@15..15
                ) |R_PAREN@16..16

